src/gui/qgslockbutton.cpp

```diff
@@ -0,0 +1,117 @@
+/***************************************************************************
+    qgslockbutton.cpp - Lock button
+     --------------------------------------
+    Date                 : July, 2017
+    Copyright            : (C) 2017 by Mathieu Pellerin
+    Email                : nirvn dot asia at gmail dot com
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgslockbutton.h"
+
+#include <QApplication>
+#include <QMouseEvent>
+#include <QPainter>
+#include <QPushButton>
+#include <QWidget>
+
+QgsLockButton::QgsLockButton( QWidget *parent )
+  : QToolButton( parent )
+  , mLocked( false )
+
+{
+  setMinimumSize( QSize( 24, 24 ) );
+  setCheckable( true );
+  connect( this, &QPushButton::clicked, this, &QgsLockButton::buttonClicked );
+}
+
+void QgsLockButton::setLocked( const bool locked )
+{
+  if ( mLocked != locked )
+    buttonClicked();
+}
+
+void QgsLockButton::buttonClicked()
+{
+  mLocked = !mLocked;
+  setDown( mLocked );
+
+  emit lockChanged( mLocked );
+
+  drawButton();
+}
+
+void QgsLockButton::changeEvent( QEvent *e )
+{
+  if ( e->type() == QEvent::EnabledChange )
+  {
+    drawButton();
+  }
+  QToolButton::changeEvent( e );
+}
+
+#if 0 // causes too many cyclical updates, but may be needed on some platforms
```

Comment:
I'd kill this. I've seen it floating around a lot in other classes, and I suspect it's just been copied and pasted around for a while now. But it's also been disabled for a long time, and I haven't seen any bug reports related.